### PR TITLE
Ensure common supertype retains nullability in inference

### LINKED ISSUE #49938: Inference doesn't find common base type for union with null or undefined - TypeScript 4.8 regression
# Bug Report

### 🔎 Search Terms

inference common base type null undefined 

### 🕗 Version & Regression Information

- This changed between versions 4.7.3 and 4.8-beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/GYVwdgxgLglg9mABAUwI4gIYBsA8AVAPgAoMAuRPAGkQCNy8BKRAb0QF8AoDrZKRAN0QBeRGBBYsAQgDcXDgHp5iAPI0AVsmiIoATwAOyAM6IA7jCgALRBDgBbWwloZDybfqNddBxACFhLRGA4OHJDKAAnGDAAc3YOL1cAEX9WIJDEMMiY6Sdw8jFbGmRwuIUlAFFw8LgSqMQAFgA6AA4KAGVrDDAAcj5gKIATazsHJATEQBlyMobGgHYKU2dEcAHkfrBkIYAfXw40TCwiQSWfamPjFbWozcQdxIYuRUQ4AGsFmGM-QFByPfRsI8QJzOgOM92mr3en1uyzAq3WNx++3+518wJO0Mu8IGDwUACpcRxELiVJZiigAB4YWx6HiGQm4+SPJQABUitnMMH4RkBsMQWHMxWwdOmlWqtSQTVaeA6EC6vUCg2G9kc4ymTya8zwiwusKuG22ohAhWKvwOAKWBSK4TROrh1wNAAZoQBGaEAJhxTwhWo+huNJURf0OKMtxRtiEQTp2rp2HtkXrePuMoZKO0x9sQgbNIaNVvDfqtGN1WIYsiZiAAglUMDpjGZLErRk4XG4DMKnqKaog6pL2p0en1FTZlWN3JNphqFiYlun9dCiCnoZkotEGABtAC6puRIMQC9zZJ2y5i6434dnNx2Kc3nqU3u7xn3-qXERXp8z2+Du6fhaPr5Pm7hteG4PAmkJ7ouf5ZKum5Fnac5ZjuSw-oeGT-jBZ4CN+kFodB75psW9oPEAA)

### 💻 Code

```ts
function equal<T>(a: T, b: T) { }

let v = null!;

// Object types with common base types

type B = { foo: string }
type D = { foo: string; bar: number }

// Error in 4.8 TS can't find common type ❌
// 4.7 T was undefined | B
equal(v as B, v as undefined | D)

// ok T is B ✅
equal(v as B, v as D)
// ok T is B | undefined ✅
equal(v as B, v as B | undefined)
```

### 🙁 Actual behavior

In 4.8 TypeScript can't find common base type between `B` and `undefined | D`. In 4.7 this used to be `B | undefined`. 

This also happens for literal types and their respective base types as well as for arrays as shown in the 'Other Examples' in the playground link. I didn't include the code as the bug is the same, it's just different instances we found in our code.

### 🙂 Expected behavior

Typescript finds the type `B | undefined` for `T` as it did in 4.7.

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #49938

Demo: [Playground Link](https://www.staging-typescript.org/play?ts=4.8.0-pr-49981-17#code/GYVwdgxgLglg9mABAUwI4gIYBsA8AVAPgAoMAuRPAGkQCNy8BKRAb0QF8AoDrZKRAN0QBeRGBBYsAQgDcXAPRzEAeRoArZNERQAngAdkAZ0QB3GFAAWiCHAC2NhLQwHkWvYa479iAELCWiYDg4cgMoACcYMABzdg5PFwARP1ZA4MRQiOjpRzDyMRsaZDDYjjRMLCJBJx9qKqNwABNkYEjkBsQAH0QEhg4FRAA9AH5S9GxKxGrG5tb2roTayaNvXv7hrjLsHGmWsDbOn2I6moElxB3Zg56+xXXN3Au9ucOJqbAm3f35xeqVm8GRqNyq9lj96u8Zk9VrcRvcQecIZ8GmCfNCARsxhVjt4Udc1rDMfCFqdfmj1kDxtiUb4uo82mSCcDsQc6ciScsGRwgA)

The common supertype of a bunch of inference candidates should include `null` and `undefined` of those candidates can be `null` or `undefined`.

#49119 and #49330 changed the way this information was stored and propagated. It turns out that the old code only worked because of a fluke related to those flags. In v4.7.4, given the types in the playground link above:

- The candidates are `B` and `D | undefined`. We declare that these are both "primary types" because neither are _exactly_ `null` or `undefined`.
- We look for the supertype of these, but because neither are supertypes of each other (because of the undefined), we end up picking `B` as it's the leftmost option.
- But then, we would rescan the input types for the types whose falsy flags include `null` or `undefined`, but using a function that is _recursive_, so we find `D | undefined`, when the intent was to just find _exactly_ `null` or `undefined`.
- We union that with `B` to get `B | D | undefined`, which reduces to `D | undefined`, and we get the supertype that everyone had expected all along, even though that's maybe not what the code was written to do!

Instead, we can just compare the candidates without `null` or `undefined`, then add `null` or `undefined` back on again if the supertype needs it; this means that the cases we supported in 4.7.4 work (and now work the same regardless of argument order).

### CODE REVIEW

**Comment by jakebailey** [comment]
On second thought, I'm not sure this is a good idea; it seems like the intent is that `getCommonSupertype` never returns any "new" types, which would mean that the original issue is not the behavior we are intending.

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test this faster
@typescript-bot pack this

**Comment by jakebailey** [comment]
Well, this at least seems less weird, but I'm not 100% certain (I'll see what happens with the above testing).

**Comment by jakebailey** [comment]
Hm, this seems to also be subtly wrong. I'll stop playing with it until we can have a discussion, I guess.

**Comment by jakebailey** [comment]
Well, I narrowed this down to operate strictly on only unions and direct types, so it shouldn't have that build problem anymore (and I encoded it as a test).

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot pack this

**Comment by jakebailey** [comment]
Demo as of this current commit: [Playground Link](https://www.staging-typescript.org/play?ts=4.8.0-pr-49981-17#code/GYVwdgxgLglg9mABAUwI4gIYBsA8AVAPgAoMAuRPAGkQCNy8BKRAb0QF8AoDrZKRAN0QBeRGBBYsAQgDcXAPRzEAeRoArZNERQAngAdkAZ0QB3GFAAWiCHAC2NhLQwHkWvYa479iAELCWiYDg4cgMoACcYMABzdg5PFwARP1ZA4MRQiOjpRzDyMRsaZDDYjjRMLCJBJx9qKqNwABNkYEjkBsQAH0QEhg4FRAA9AH5S9GxKxGrG5tb2roTayaNvXv7hrjLsHGmWsDbOn2I6moElxB3Zg56+xXXN3Au9ucOJqbAm3f35xeqVm8GRqNyq9lj96u8Zk9VrcRvcQecIZ8GmCfNCARsxhVjt4Udc1rDMfCFqdfmj1kDxtiUb4uo82mSCcDsQc6ciScsGRwgA)

**Comment by jakebailey** [comment]
RE: the RWC tests, I think there might be one error in there that's real, but I'm having trouble understanding what's going on there because it looks like a bunch of unrelated things are showing up in the diff that shouldn't be there if the diff was running on the same version of the test codebases.

**Comment by jakebailey** [comment]
@typescript-bot perf test this faster

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21218,19 +21218,31 @@ namespace ts {
...
+            if (shouldHaveNull && !someHas(supertypeOrUnion, TypeFlags.Null)) {
+                missingNullableFlags |= TypeFlags.Null;
+            }
+            if (shouldHaveUndefined && !someHas(supertypeOrUnion, TypeFlags.Undefined)) {
+                missingNullableFlags |= TypeFlags.Undefined;
+            }
+            return getNullableType(supertypeOrUnion, missingNullableFlags);
+
+            function someHas(type: Type, flags: TypeFlags): boolean {
+                return someType(type, t => !!(t.flags & flags));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:21244]: This feels common enough that I would rather us lift it up to `someTypeHasFlags`

Comment (jakebailey) [reply src/compiler/checker.ts:21244]: I can do that, yeah, though I think most places just entirely inline it, so I could just do that too.

**Comment by jakebailey** [comment]
#50021 is a more complete fix than mine now, so, closing.

**Pull request closed without merging**